### PR TITLE
fix: use async neo4j

### LINKED ISSUE #166: Adding a children blocks the API server
### What
If we add a children to a node the server, it makes the server unresponsive (infinite loop ?)

Related to:
- #181

Blocks:
- #162

### LINKED ISSUE #181: Neo4J crashing when adding a child
### What
We have a problem experimented in staging where neo4j crashes without a log.

It happens when adding a child to an entry.
You have to remove and recreate the docker container for it to start again (just restarting is not enough)

### Steps to reproduce the behavior:
1. Go to an entry in a big enough taxonomy (like allergens)
2. go to root nodes and choose a node to edit it
3. click on the button to add a child
4. application is not responding any more.
5. behind the scene:
   * neo4j wont respond to any query anymore 
   * even a restart does not make it back (maybe because of some lock, but again neo4j logs are mute)
   * you have to remove / recreate the container

### Expected behavior
This should not crash :grimacing:

### PR DESCRIPTION
**Pull request by alexgarel**
### What

There are 3 commits in  this PR (better review them individually):

* build: handle neo4j memory configration
  + upgrade neo4j and some libs
  + enable reload in dev mode
   + avoid test failing because of neo4j running

* fix: fix bugs in parser
  * tracing errors did raise an exception if there were parameters (added a test)
  * better handling of is_before (avoid class attribute)

* fix: use async neo4j in entries.py  
  Related issues:
  - Fixes #181
  - Fixes #166 

The last one is a big one. I tested around and it seems ok, but I can't be sure I didn't break anything.

We **really need** to add tests to this project !

### CODE REVIEW

File: backend/editor/entries.py [hunk 1]
```diff
@@ -483,29 +493,27 @@ def update_node_children(self, entry, new_children_ids):
...
             main_language_code = child.split(":", 1)[0]
-            created_node_id = self.create_node("ENTRY", child, main_language_code)
+            created_node_id = await self.create_node("ENTRY", child, main_language_code)
             created_child_ids.append(created_node_id)
 
             # TODO: We would prefer to add the node just after its parent entry
-            self.add_node_to_end("ENTRY", created_node_id)
+            await self.add_node_to_end("ENTRY", created_node_id)
 
         # Stores result of last query executed
```
Comment (aadarsh-ram) [inline backend/editor/entries.py:518]: I think we forget to create relations if the nodes are already existing. So, the function from this line, needs to be changed as follows:
```
        children_to_create = created_child_ids + existing_ids
        # Stores result of last query executed
        result = []
        for child in children_to_create:
            # Create new relationships if it doesn't exist
            query = f"""
                MATCH (parent:{self.project_name}:ENTRY), (new_child:{self.project_name}:ENTRY)
                WHERE parent.id = $id AND new_child.id = $child
                MERGE (new_child)-[r:is_child_of]->(parent)
            """
            _result = await get_current_transaction().run(query, {"id": entry, "child": child})
            result = list(await _result.value())

        return result
```

File: backend/editor/api.py [hunk 2]
```diff
@@ -149,8 +151,8 @@ async def list_all_projects(response: Response, status: Optional[StatusFilter] =
     List projects created in the Taxonomy Editor with a status filter
     """
     # Listing all projects doesn't require a taxonomy name or branch name
-    taxonony = TaxonomyGraph("", "")
-    result = list(taxonony.list_projects(status))
+    taxonomy = TaxonomyGraph("", "")
+    result = taxonomy.list_projects(status)
```
Comment (aadarsh-ram) [inline backend/editor/api.py:155]:
```suggestion
    result = await taxonomy.list_projects(status)
```

File: backend/editor/api.py [hunk 3]
```diff
@@ -409,8 +400,10 @@ async def edit_entry(request: Request, branch: str, taxonomy_name: str, entry: s
     """
     taxonomy = TaxonomyGraph(branch, taxonomy_name)
     incoming_data = await request.json()
-    result = taxonomy.update_nodes("ENTRY", entry, incoming_data)
-    updated_entry = list(result)
+    log.info("start entry update")
+    updated_entry = await taxonomy.update_nodes("ENTRY", entry, incoming_data)
+    log.info("end entry update")
+    log.info(f"return {updated_entry}")
```
Comment (aadarsh-ram) [inline backend/editor/api.py:406]: Can we remove this log?

File: backend/editor/api.py [hunk 4]
```diff
@@ -423,8 +416,10 @@ async def edit_entry_children(request: Request, branch: str, taxonomy_name: str,
     """
     taxonomy = TaxonomyGraph(branch, taxonomy_name)
     incoming_data = await request.json()
-    result = taxonomy.update_node_children(entry, incoming_data)
-    updated_children = list(result)
+    log.info("start children update")
+    updated_children = await taxonomy.update_node_children(entry, incoming_data)
+    log.info("end children update")
+    log.info(f"return children {updated_children}")
```
Comment (aadarsh-ram) [inline backend/editor/api.py:422]: Can we remove this too?

**Review by aadarsh-ram — CHANGES_REQUESTED** [review]
Thanks for the amazing work! I requested some changes, where the code's not working due to the changes to async are breaking. Please do check them out.

File: backend/editor/api.py (hunk 3 above)
Comment (alexgarel) [reply backend/editor/api.py:406]: yes totally ! I added it when I was debugging !

Comment (alexgarel) [inline backend/editor/api.py:406]:
```suggestion
    updated_entry = await taxonomy.update_nodes("ENTRY", entry, incoming_data)
```

File: backend/editor/api.py (hunk 4 above)
Comment (alexgarel) [reply backend/editor/api.py:422]: yes again :-)

Comment (alexgarel) [inline backend/editor/api.py:422]:
```suggestion
    updated_children = await taxonomy.update_node_children(entry, incoming_data)
```

File: backend/editor/graph_db.py [hunk 5]
```diff
@@ -3,34 +3,48 @@
...
+
+
 txn = contextvars.ContextVar("txn")
 txn.set(None)
 
 session = contextvars.ContextVar("session")
 session.set(None)
 
+test = contextvars.ContextVar("test")
+test.set(None)
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply backend/editor/graph_db.py:24]: Haha I though I had removed them :sweat_smile: 

I will do !

File: parser/openfoodfacts_taxonomy_parser/parser.py [hunk 6]
```diff
@@ -369,7 +373,7 @@ def create_nodes(self, filename, multi_label):
 
     def create_previous_link(self, multi_label):
         self.parser_logger.info("Creating 'is_before' links")
-        query = f"MATCH(n:{multi_label}) WHERE exists(n.is_before) return n.id, n.is_before"
+        query = f"MATCH(n:{multi_label}) WHERE n.is_before IS NOT NULL return n.id, n.is_before"
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply parser/openfoodfacts_taxonomy_parser/parser.py:376]: @aadarsh-ram we need to store it temporarily in node, but we could remove it just after link creations (I though it was already the case). We can open a ticket for that, but it's not a big deal if it remains (but it makes a good first issue :-) )

File: backend/editor/entries.py (hunk 1 above)
Comment (alexgarel) [reply backend/editor/entries.py:518]: Cool we also fix bugs while reviewing :-)

File: backend/editor/graph_db.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply backend/editor/graph_db.py:24]: At some point as was not sure if our transaction mechanism was working as expected, that's why I added this… and it works very well :-)

File: parser/openfoodfacts_taxonomy_parser/parser.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply parser/openfoodfacts_taxonomy_parser/parser.py:376]: https://github.com/openfoodfacts/taxonomy-editor/issues/183

**Comment by alexgarel** [comment]
@aadarsh-ram modifications done.

File: backend/editor/graph_db.py [hunk 7]
```diff
@@ -3,34 +3,41 @@
...
+import logging
 
-from neo4j import GraphDatabase  # Interface with Neo4J
+import neo4j  # Interface with Neo4J
 
 from . import settings  # Neo4J settings
 from .exceptions import SessionMissingError  # Custom exceptions
 from .exceptions import TransactionMissingError
 
+log = logging.getLogger(__name__)
```
Comment (aadarsh-ram) [inline backend/editor/graph_db.py:14]: I think we don't use this anywhere in this file, so I think this line and the imports can be removed

File: backend/requirements.txt [hunk 8]
```diff
@@ -1,20 +1,20 @@
 anyio==3.6.1
 click==8.1.3
-fastapi==0.79.0
+fastapi==0.89.1
```
Comment (aadarsh-ram) [inline backend/requirements.txt:3]: @alexgarel Hopefully all the necessary dependencies for fastAPI and other packages has also been updated.

File: docker/dev.yml [hunk 9]
```diff
@@ -20,6 +20,9 @@ services:
       # for linting / checks purpose
       - ./backend/pyproject.toml:/code/pyproject.toml
       - ./backend/setup.cfg:/code/setup.cfg
+    # uvicorn with live reload
+    command: ["uvicorn", "editor.api:app", "--host", "0.0.0.0", "--port", "80", "--reload", "--reload-dir", "/parser", "--reload-dir", "/code/editor"]
```
Comment (aadarsh-ram) [inline docker/dev.yml:24]: Will be useful for dev! :)

**Review by aadarsh-ram — APPROVED** [review]
@alexgarel I'll approve. Feel free to merge after adding any necessary changes,

File: backend/editor/graph_db.py (hunk 7 above)
Comment (alexgarel) [reply backend/editor/graph_db.py:14]: Yep, but we should log more (at least debug or info) so I think we can keep it (it's a good pattern to have it)

File: backend/requirements.txt (hunk 8 above)
Comment (alexgarel) [reply backend/requirements.txt:3]: seems so :-)

File: docker/dev.yml (hunk 9 above)
Comment (alexgarel) [reply docker/dev.yml:24]: Yes I though it was already done !

**Pull request merged**